File: src/main/java/Tries/TrieII.java
package main.java.Tries;

import java.util.HashMap;
import java.util.Map;

class TrieIINode {
    private final Map<Character, TrieIINode> map;
    public int prefixCount;
    public int endCount;

    public TrieIINode() {
        this.map = new HashMap<>();
        this.prefixCount = 0;
        this.endCount = 0;
    }

    public TrieIINode getOrModify(char c) {
        TrieIINode next = map.computeIfAbsent(c, k -> new TrieIINode());
        next.prefixCount++;
        return next;
    }

    public TrieIINode get(char c) {
        return map.get(c);
    }
}

public class TrieII {
    private final TrieIINode root;

    public TrieII() {
        this.root = new TrieIINode();
    }

    public void insert(String word) {
        TrieIINode currNode = root;

        for (int i = 0; i < word.length(); i++) {
            currNode = currNode.getOrModify(word.charAt(i));
        }

        currNode.endCount++;
    }

    public int countWordEqualTo(String word) {
        TrieIINode currNode = root;

        for (int i = 0; i < word.length() && currNode != null; i++) {
            currNode = currNode.get(word.charAt(i));
        }

        return currNode == null ? 0 : currNode.endCount;
    }

    public int countStartingWith(String prefix) {
        TrieIINode currNode = root;

        for (int i = 0; i < prefix.length() && currNode != null; i++) {
            currNode = currNode.get(prefix.charAt(i));
        }

        return currNode == null ? 0 : currNode.prefixCount;
    }

    public void erase(String word) {
        TrieIINode currNode = root;
        TrieIINode[] visited = new TrieIINode[word.length()];

        for (int i = 0; i < word.length(); i++) {
            currNode = currNode.get(word.charAt(i));

            if (currNode == null) {
                for (int j = 0; j < i; j++) {
                    visited[j].prefixCount++;
                }
                return;
            }

            currNode.prefixCount--;
            visited[i] = currNode;
        }

        if (currNode.endCount == 0) {
            for (TrieIINode node : visited) {
                node.prefixCount++;
            }
            return;
        }

        currNode.endCount--;
    }
}

File: src/main/python/Tries/trieII.py
from __future__ import annotations

class TrieNode:
    def __init__(self) -> None:
        self.map: dict[str, TrieNode] = {}
        self.prefix_count = 0
        self.end_count = 0

    def get(self, char: str) -> TrieNode | None:
        return self.map.get(char)

    def get_or_modify(self, char: str) -> TrieNode:
        if char not in self.map:
            self.map[char] = TrieNode()

        self.map[char].prefix_count += 1
        return self.map[char]


class TrieII:
    def __init__(self) -> None:
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        curr_node = self.root

        for char in word:
            curr_node = curr_node.get_or_modify(char)

        curr_node.end_count += 1

    def count_word_equal_to(self, word: str) -> int:
        curr_node = self.root

        for char in word:
            if curr_node is None:
                return 0
            curr_node = curr_node.get(char)

        return 0 if curr_node is None else curr_node.end_count

    def count_starting_with(self, prefix: str) -> int:
        curr_node = self.root

        for char in prefix:
            if curr_node is None:
                return 0
            curr_node = curr_node.get(char)

        return 0 if curr_node is None else curr_node.prefix_count

    def erase(self, word: str) -> None:
        if self.count_word_equal_to(word) == 0:
            return

        curr_node = self.root

        for char in word:
            curr_node = curr_node.get(char)
            curr_node.prefix_count -= 1

        curr_node.end_count -= 1